pacman/data/pacman_data_view.py

```diff
@@ -0,0 +1,512 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+from spinn_utilities.log import FormatAdapter
+from spinn_machine.data import MachineDataView
+from pacman.exceptions import PacmanNotPlacedError
+from pacman.model.graphs.application import ApplicationGraph
+
+logger = FormatAdapter(logging.getLogger(__name__))
+# pylint: disable=protected-access
+
+
+class _PacmanDataModel(object):
+    """
+    Singleton data model
+
+    This class should not be accessed directly please use the DataView and
+    DataWriter classes.
+    Accessing or editing the data held here directly is NOT SUPPORTED
+
+    There may be other DataModel classes which sit next to this one and hold
+    additional data. The DataView and DataWriter classes will combine these
+    as needed.
+
+    What data is held where and how can change without notice.
+    """
+
+    __singleton = None
+
+    __slots__ = [
+        # Data values cached
+        "_graph",
+        "_placements",
+        "_plan_n_timesteps",
+        "_precompressed",
+        "_routing_infos",
+        "_routing_table_by_partition",
+        "_runtime_graph",
+        "_tags",
+        "_uncompressed",
+        "_vertices_or_edges_added",
+    ]
+
+    def __new__(cls):
+        if cls.__singleton:
+            return cls.__singleton
+        # pylint: disable=protected-access
+        obj = object.__new__(cls)
+        cls.__singleton = obj
+        obj._clear()
+        return obj
+
+    def _clear(self):
+        """
+        Clears out all data
+        """
+        self._graph = None
+        # set at the start of every run
+        self._plan_n_timesteps = None
+        self._vertices_or_edges_added = False
+        self._hard_reset()
+
+    def _hard_reset(self):
+        """
+        Clears out all data that should change after a reset and graph change
+        """
+        if self._graph:
+            self._graph.reset()
+        self._placements = None
+        self._precompressed = None
+        self._uncompressed = None
+        self._runtime_graph = None
+        self._routing_infos = None
+        self._routing_table_by_partition = None
+        self._tags = None
+        self._soft_reset()
+
+    def _soft_reset(self):
+        """
+        Clears timing and other data that should changed every reset
+        """
+        # Holder for any later additions
+
+
+class PacmanDataView(MachineDataView):
+    """
+    Adds the extra Methods to the View for pacman level.
+
+    See UtilsDataView for a more detailed description.
+
+    This class is designed to only be used directly within the PACMAN
+    repository as all methods are available to subclasses
+    """
+
+    __pacman_data = _PacmanDataModel()
+    __slots__ = []
+
+    # graph methods
+
+    @classmethod
+    def has_application_vertices(cls):
+        """
+        Reports if the user level graph has application vertices.
+
+        Semantic sugar for get_graph().n_vertices
```

Comment:
I suppose strictly speaking it's semantic sugar for "if get_graph().n_vertices > 0"... ?  Especially as later you actually have get_n_vertices as a function.

Reply:
https://github.com/SpiNNakerManchester/PACMAN/issues/447

Hangover from when we had machine graphs. Flaged for removal in round 2

Reply:
done